kernel/posix/mqueue.c

```diff
@@ -0,0 +1,532 @@
+		msg_queue->prev = NULL;
+
+		if (mq_list != NULL) {
+			mq_list->prev = msg_queue;
+		}
+
+		msg_queue->next = mq_list;
+		mq_list = msg_queue;
+	} else {
+		k_sem_take(&mq_sem, K_FOREVER);
+
+		if (msg_queue->ref_count >= CONFIG_MAX_DESC_PERMQ) {
+			k_sem_give(&mq_sem);
+			errno = ENOSPC;
+			k_mem_pool_free(&block_memq_desc);
+			return (mqd_t)mqd;
+		}
+
+		msg_queue->ref_count += 1;
+	}
+
+	k_sem_give(&mq_sem);
+	msg_queue_desc->mqueue = msg_queue;
+	msg_queue_desc->flags = (oflags & O_NONBLOCK) ? O_NONBLOCK : 0;
+
+done:
+	if (error) {
+		errno = ENOSPC;
+		if (msg_queue != NULL) {
+			if (msg_queue->block_buffer_name != NULL) {
+				k_mem_pool_free(&block_track);
+			}
+
+			if (msg_queue->name != NULL) {
+				k_mem_pool_free(&block_memq_name);
+			}
+
+			k_mem_pool_free(&block_memq_obj);
+		}
+
+		if (msg_queue_desc != NULL) {
+			k_mem_pool_free(&block_memq_desc);
+		}
+	} else {
+		mqd = msg_queue_desc;
+	}
+
+	return (mqd_t)mqd;
+}
+
+/**
+ * @brief Close a message queue descriptor.
+ *
+ * See IEEE 1003.1
+ */
+int mq_close(mqd_t mqdes)
+{
+	mqueue_desc *mqd = (mqueue_desc *)mqdes;
+	mqueue_object  *msg_queue;
+	struct k_mem_block *block;
+
+	if (mqd == NULL) {
+		errno = EBADF;
+		return -1;
+	}
+
+	msg_queue = mqd->mqueue;
+	block = &mqd->block;
+	k_mem_pool_free(block);
+	k_sem_take(&mq_sem, K_FOREVER);
+	msg_queue->ref_count -= 1;
+	k_sem_give(&mq_sem);
+	return 0;
+}
+
+/**
+ * @brief Remove a message queue.
+ *
+ * See IEEE 1003.1
+ */
+int mq_unlink(const char *name)
+{
+	mqueue_object  *msg_queue;
+	mqueue_object  *next_mqueue, *prev_mqueue;
+	struct k_mem_block *block;
+
+	k_sem_take(&mq_sem, K_FOREVER);
+	msg_queue = find_in_list(name);
+
+	if (msg_queue != NULL && msg_queue->ref_count == 0) {
+		/* If he message queue is in the list, remove it. */
+		if (mq_list == msg_queue) {
+			mq_list = msg_queue->next;
+		} else {
+			prev_mqueue = msg_queue->prev;
+			next_mqueue = msg_queue->next;
+
+			if (prev_mqueue) {
+				prev_mqueue->next = next_mqueue;
+			}
+
+			if (next_mqueue) {
+				next_mqueue->prev = prev_mqueue;
+			}
+		}
+
+		k_msgq_purge(&msg_queue->queue);
+		msg_queue->next = msg_queue->prev = NULL;
+		k_sem_give(&mq_sem);
+
+		/* Free up name */
+		block = &msg_queue->block_buffer_name->block_name;
+		k_mem_pool_free(block);
+
+		/* Free up buffer track */
+		block = &msg_queue->block_buffer_name->block_buffer;
+		k_mem_pool_free(block);
+
+		/* Free up Queue object */
+		block = &msg_queue->block_mqueue;
+		k_mem_pool_free(block);
+
+		return 0;
+	}
+
+	k_sem_give(&mq_sem);
+	errno = EBADF;
+	return -1;
+}
+
+/**
+ * @brief Send a message to a message queue.
+ *
+ * See IEEE 1003.1
+ */
+int mq_send(mqd_t mqdes, const char *msg_ptr, size_t msg_len,
+	    unsigned int msg_prio)
+{
+	mqueue_desc *mqd = (mqueue_desc *)mqdes;
+	mqueue_object  *msg_queue;
+	s32_t  timeout = K_FOREVER, ret = -1;
+
+	if (mqd == NULL) {
+		errno = EBADF;
+		return ret;
+	}
+
+	msg_queue = mqd->mqueue;
```

Comment:
Seems like a silly variable to have just so you can avoid writing "mqd->mqueue->queue" below.